scheduler/test/tbf/test/ratelimit.clj

```diff
@@ -0,0 +1,92 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns tbf.test.ratelimit
+  (:use clojure.test)
+  (:require [tbf.core]
+            [tbf.ratelimit]))
+
+(deftest independent-keys-1
+  (let [ratelimit (tbf.ratelimit/tbf-ratelimiter-make 60000 60 10)]
+    (tbf.ratelimit/earn-tokens! ratelimit "Foo1")
+    (tbf.ratelimit/time-until ratelimit "Foo2")
+    (tbf.ratelimit/in-debt? ratelimit "Foo3")
+    (tbf.ratelimit/spend-tokens! ratelimit "Foo4" 100)
+    (is (= 4 (.size (.asMap (:cache ratelimit)))))))
+
+(deftest independent-keys-2
+  (let [ratelimit (tbf.ratelimit/tbf-ratelimiter-make 60000 60 10)]
+    (tbf.ratelimit/earn-tokens! ratelimit "Foo4")
+    (tbf.ratelimit/earn-tokens! ratelimit "Foo1")
+    (tbf.ratelimit/time-until ratelimit "Foo1")
+    (tbf.ratelimit/time-until ratelimit "Foo2")
+    (tbf.ratelimit/in-debt? ratelimit "Foo2")
+    (tbf.ratelimit/in-debt? ratelimit "Foo3")
+    (tbf.ratelimit/spend-tokens! ratelimit "Foo3" 100)
+    (tbf.ratelimit/spend-tokens! ratelimit "Foo4" 100)
+    (is (= 4 (.size (.asMap (:cache ratelimit)))))))
+
+(deftest in-debt?-1
+  (let [ratelimit (tbf.ratelimit/tbf-ratelimiter-make 60 60000 10)]
+    (is (not (tbf.ratelimit/in-debt? ratelimit "Foo")))
+    ;; take away the full bucket it starts with... (60 tokens)
+    (is (= 0 (tbf.ratelimit/spend-tokens! ratelimit "Foo" 60)))
+    ;; Should be able to do this first request almost instantly.
+    (is (= 0 (tbf.ratelimit/spend-tokens! ratelimit "Foo" 60000)))
+    ;; Second request should need to sleep for 60 seconds.
+    (is (< 55000 (tbf.ratelimit/spend-tokens! ratelimit "Foo" 100) 65000))
+    (is (= 1 (.size (.asMap (:cache ratelimit)))))
+    (is (not (tbf.ratelimit/in-debt? ratelimit "Bar")))
+    (is (tbf.ratelimit/in-debt? ratelimit "Foo"))))
+
+(deftest earning-tokens-explicit
+  (let [ratelimit (tbf.ratelimit/tbf-ratelimiter-make 20 60000 10)]
+    ;; take away the full bucket it starts with... (20 tokens)
+    (is (= 0 (tbf.ratelimit/spend-tokens! ratelimit "Foo1" 20)))
+    (is (= 0 (tbf.ratelimit/spend-tokens! ratelimit "Foo2" 20)))
```

Comment:
Why is safe to assume that enought time will not have elapsed that will trigger tokens to be refilled?

Reply:
Not sure what you're asking here. The logic is we always start with a full bucket. To make it easier to unit test, I always remove the bucket entirely, so the balance is just at zero. Tokens refill continously, so there might be a token here. All of the later checks use a range.